src/ReactGraph/DependencyEngine.cs

```
@@ -8,46 +8,35 @@ namespace ReactGraph
 {
     public class DependencyEngine
     {
-        private readonly List<INotificationStrategy> notificationStrategies;
-        private readonly Dictionary<Tuple<object, string>, DependencyInfo> nodeLookup = new Dictionary<Tuple<object, string>, DependencyInfo>();
-        private readonly Dictionary<DependencyInfo, object> instancesToSwitch = new Dictionary<DependencyInfo, object>();
-        private readonly Dictionary<DependencyInfo, object> leafDependencies = new Dictionary<DependencyInfo, object>();
-        private readonly List<object> instancesBeingTracked = new List<object>();
-        private readonly DirectedGraph<DependencyInfo> graph;
+        private readonly DirectedGraph<INodeInfo> graph;
         private readonly ExpressionParser expressionParser;
+        private NodeRepository nodeRepository;
         private bool isExecuting;
 
         public DependencyEngine()
         {
-            notificationStrategies = new List<INotificationStrategy>
-            {
-                new NotifyPropertyChangedStrategy(this)
-            };
-            graph = new DirectedGraph<DependencyInfo>();
-            expressionParser = new ExpressionParser();
+            graph = new DirectedGraph<INodeInfo>();
+            nodeRepository = new NodeRepository(this);
+            expressionParser = new ExpressionParser(nodeRepository);
         }
 
         public event Action<object, string> SettingValue = (o, s) => { };
 
-        public void PropertyChanged(object instance, string property)
+        public void ValueHasChanged(object instance, string key)
         {
-            var key = Tuple.Create(instance, property);
-            if (!nodeLookup.ContainsKey(key)) return;
-            var node = nodeLookup[key];
-
-            SwitchTransientInstances(node);
-            UpdateLeafDependencies(node);
-
-
+            if (!nodeRepository.Contains(instance, key)) return;
```

Throw instead of returning silently?